docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+:::
+
+## Optional Params
+Params can be made optional by adding a `?` before the name. The `year` param is not optional and will have the type `string` when being accessed. But the `month` param is optional and will have the type `string | undefined` when being accessed.
+
+```ts {3}
+const events = createRoute({
+  name: 'events',
+  path: '/events/[year]/[?month]',
+})
+```
+
+## Param Types
+By default all params are strings. But you can define a specific type for a param by using the `path` or `query` utilities. Now the `year` param will have the type `number` when being accessed.
+
+```ts {1,5-7}
+import { path } from '@kitbag/router'
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+  }),
+})
+```
+
+:::info Route Matching
+When using a param type, if the url cannot be parsed as the specified type it will not be matched. For example if the url was `/events/two-thousand-and-twenty-four/september` it would not be matched because `two-thousand-and-twenty-four` is not a number.
+:::
+
+### Built-in Param Types
+Kitbag Router comes with a few built-in param types:
+
+| Type     | Default | Description                                                                                   |
+| -------- | ------- | --------------------------------------------------------------------------------------------- |
+| `String` | ✓       | Any string value                                                                              |
+| `Number` |         | Any value that can be parsed with the `Number` constructor                                    |
+| `Boolean`|         | The literal values `"true"` or `"false"`                                                      |
+| `Date`   |         | Any value that can be parsed with the `Date` constructor. Uses `toISOString` when serializing |
+| `RegExp` |         | A literal regex expression                                                                    |
+| `JSON`   |         | Any value that can be parsed with the `JSON.parse` method                                     |
+
+## Custom Param Types
+You can define your own param types using the `createParam` utility. This serves three purposes:
+1. Define the type of the param.
+2. Define how the param should be serialized and deserialized.
+3. Determine if a value is valid.
+
+Using custom param types you can define more complex validation rules. For example, this month param must be a valid month name. If it is not, the route will not match.
+
+```ts {3-11,17}
+import { createParam, createRoute, path } from '@kitbag/router'
+
+const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december'] as const
+
+const monthParam = createParam((value, { invalid }) => {
+  if (months.includes(value as any)) {
+    return value
+  }
+
+  invalid(`Invalid month: ${value}`)
+})
```

Comment:
the way this is defined makes sense if you want your value to be `january | february | ...` but as its written you'd still have `string` since you're `if(months.includes...)` is not a type guard.

I prefer either 1. take this suggestion and let the param type be `string`. or 2 maybe just "import" a type guard and assume devs know what type guards do

```suggestion
const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december']

const monthParam = createParam((value, { invalid }) => {
  if (months.includes(value)) {
    return value
  }

  invalid(`Invalid month: ${value}`)
})
```

Reply:
also, don't we want to `throw invalid(...)`? otherwise our param also appears to return possibly `undefined`

Reply:
```suggestion
const months = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december']

const monthParam = createParam((value, { invalid }) => {
  if (months.includes(value)) {
    return value
  }

  throw invalid(`Invalid month: ${value}`)
})
```